[feature](filecache)A 2Q-LRU mechanism for protecting hotspot data in the normal queue

wenzhenghu opened the pull request:
# PR Objective

In this PR, We have introduced a segmented（hot-cold） LRU mechanism To make sure that **frequently accessed (“hot”) pages** remain in the filecache, even as read-ahead and full table/index scans bring in new blocks that might or might not be accessed afterward. 

# Implementation Description
Specifically, we split the NORMAL queue of the filecache into two LRU queues, which cache hot/young data and cold/old data respectively. （The design of the hot-cold LRU mechanism references the implementation of MySQL InnoDB's midpoint LRU. For details, please refer to the [link.](https://dev.mysql.com/doc/refman/8.4/en/innodb-performance-midpoint_insertion.html)）

1.  The hot data remains in the original NORMAL queue. **A new cold data queue, COLD_NORMAL, is added**. Data that originally entered the NORMAL queue will now first enter the COLD_NORMAL queue and can only be promoted to the NORMAL queue after meeting certain conditions（reside in COLD_NORMAL queue more than normal_queue_cold_time_ms before being accessed again）.

2.  During cache block eviction, the newly added **COLD_NORMAL queue has an eviction priority between the DISPOSABLE queue and the NORMAL queue**. It is important to note that cache blocks evicted from the NORMAL queue will not enter the COLD_NORMAL queue.

## BE-Side Configuration Parameter
In terms of implementation, we have added three new BE configuration parameters:
### enable_file_cache_normal_queue_2qlru (Boolean)
- **Default**: True. 
- Controls whether to enable the file cache 2-queue（hot-cold） LRU mechanism in normal queue.
### file_cache_2qlru_cold_blocks_percent (Double)
- **Default**: 20. 
- Controls the target percentage of total normal queue space allocated to the COLD_NORMAL queue.
### file_cache_2qlru_cold_blocks_promotion_ms (Integer)
- **Default**: 1000. 
- Defines the minimum time a cache block must reside in the COLD_NORMAL queue before being accessed when it becomes eligible for promotion to the NORMAL queue.

## Core Function Modifications​
| Function | Purpose | Modification |
|--------|--------|--------|
| **add_cell()** | Creates a new cache block. | Adds a special check: upon the first access of a NORMAL-type block, it is converted to a COLD_NORMAL-type and inserted into the corresponding queue. This initializes the block as "cold" data. |
| **use_cell()** | Handles access to an existing cache block. | Implements the logic to promote a COLD_NORMAL-type block converted to a NORMAL-type block upon access, moving it to the appropriate "hot" queue. |
| **evict_from_other_queue()** | Handles cache eviction to free up space. | By adjusting the queue priority order to DISPOSABLE < COLD_NORMAL < NORMAL < INDEX, the evict_from_other_queue() function ensures that lower-priority data is evicted before higher-priority data, implementing the "evict cold, retain hot" strategy. |

## TPCH Test Results
The test results indicate that after enabling the cold-hot dual LRU feature, both the cache hit rate and query performance have improved, with a more significant performance improvement observed under low cache allocation ratios. The testing utilized the TPC-H benchmark. The result data is as follows:

### TPCH 100 (about 25GB data should cache)
| Cache Configuration   |   Single LRU Hit Rate (%) |   Single LRU Performance (Relative) |   Dual LRU Hit Rate (%) |   Dual LRU Performance (Relative) |
|:----------------------|--------------------------:|------------------------------------:|------------------------:|----------------------------------:|
| 20G                   |                     92.74 |                                 1 |                   96.11 |                            1.04 |
| 15G                   |                     73.32 |                                 1 |                   82.33 |                            1.12 |
| 10G                   |                     40.23 |                                 1 |                   62.06 |                            1.54 |

## Compatibility-Related Implementation:
-  When dumping with a single LRU queue and loading with a dual LRU queue:​ In this scenario, all the loaded data enters the NORMAL queue. Because the available cache space for itself is reduced, the NORMAL queue will (according to the existing mechanism of the filecache) borrow space from the COLD_NORMAL lqueue. 
-  When dumping with a dual LRU queue and loading with a single LRU queue:​ It is necessary to add the data from the COLD_NORMAL queue to the tail of the NORMAL queue according to the LRU order. 

Related Old PR: https://github.com/apache/doris/pull/57353 , it is old draft pr

None

### Check List (For Author)

- Test <!-- At least one of them must be included. -->
    - [x] Regression test
    - [x] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason <!-- Add your reason?  -->

- Behavior changed:
    - [ ] No.
    - [ ] Yes. <!-- Explain the behavior change -->

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. <!-- Add document PR link here. eg: https://github.com/apache/doris-website/pull/1214 -->

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label <!-- Add branch pick label that this PR should merge into -->

xuchenhao committed:
Build the framework for NORMAL queue hot-cold data separation.

xuchenhao committed:
Fixed the bug where metrics are displaying incorrectly.

xuchenhao committed:
Added the hits_ratio_1m metric.

xuchenhao committed:
Add a unit test case.

xuchenhao committed:
Support cold-hot separation for try_reserve_from_other_queue.

xuchenhao committed:
Update .idea/vcs.xml

xuchenhao committed:
Merge remote-tracking branch 'origin/master' into normal_queue_cold_hot_seperation

xuchenhao committed:
Update vcs.xml

xuchenhao committed:
Format the code.

xuchenhao committed:
Merge remote-tracking branch 'origin/normal_queue_cold_hot_seperation' into normal_queue_cold_hot_seperation

xuchenhao committed:
Format the code.

xuchenhao committed:
Format the code.

xuchenhao committed:
Update config.cpp.

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_seperation

xuchenhao committed:
Update vcs.xml

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_seperation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_seperation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_seperation

xuchenhao committed:
Update ut cases.

xuchenhao committed:
Update rt cases.

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_seperation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Format the code.

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

wenzhenghu committed:
Merge branch 'master' into normal_queue_cold_hot_separation

dataroaring commented:
2Q lru is more general name?

xuchenhao committed:
Add config normal_queue_cold_time_ms.

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge remote-tracking branch 'origin/normal_queue_cold_hot_separation' into normal_queue_cold_hot_separation

xuchenhao committed:
Format the code.

wenzhenghu commented:
> 2Q lru is more general name?

agree!

xuchenhao committed:
Update ut cases.

xuchenhao committed:
Format the code.

xuchenhao committed:
Update config name.

xuchenhao committed:
Format the code.

xuchenhao commented:
run buildall

xuchenhao committed:
Update rt cases.

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

wenzhenghu committed:
Merge branch 'master' into normal_queue_cold_hot_separation

wenzhenghu committed:
Merge branch 'master' into normal_queue_cold_hot_separation

wenzhenghu commented:
run buildall

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `92.48% (209/226)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57410_e756b9c3800db362d730611df9db2c4c83a6ad31_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57410_e756b9c3800db362d730611df9db2c4c83a6ad31_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 72.43% (24418/33713) |
| Line Coverage     | 59.27% (256758/433218) |
| Region Coverage   | 54.93% (215845/392950) |
| Branch Coverage   | 56.46% (92342/163561) |

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Format the code.

xuchenhao commented:
run buildall

xuchenhao committed:
Update fs_file_cache_storage.cpp.

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Add new FileCacheType.

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao commented:
run buildall

xuchenhao committed:
Add new FileCacheType.

xuchenhao committed:
Merge remote-tracking branch 'origin/normal_queue_cold_hot_separation' into normal_queue_cold_hot_separation

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/39e60670fd814ea1043b594111f9689d47c63a7d_39e60670fd814ea1043b594111f9689d47c63a7d_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/39e60670fd814ea1043b594111f9689d47c63a7d_39e60670fd814ea1043b594111f9689d47c63a7d_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 80.76% (1675/2074) |
| Line Coverage     | 66.96% (29520/44083) |
| Region Coverage   | 67.59% (14699/21747) |
| Branch Coverage   | 57.86% (7822/13518) |

xuchenhao committed:
Add new FileCacheType.

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `79.65% (180/226)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/2c597f9121711cfa1a2ca2a5e449c423c14e1a2e_2c597f9121711cfa1a2ca2a5e449c423c14e1a2e/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/2c597f9121711cfa1a2ca2a5e449c423c14e1a2e_2c597f9121711cfa1a2ca2a5e449c423c14e1a2e/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 52.77% (18225/34534) |
| Line Coverage     | 38.14% (165869/434864) |
| Region Coverage   | 33.12% (128922/389230) |
| Branch Coverage   | 33.88% (55343/163370) |

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/0fd26e6ce2bd72e716396ca3eba0125648eb78cc_0fd26e6ce2bd72e716396ca3eba0125648eb78cc_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/0fd26e6ce2bd72e716396ca3eba0125648eb78cc_0fd26e6ce2bd72e716396ca3eba0125648eb78cc_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 80.76% (1675/2074) |
| Line Coverage     | 66.96% (29520/44083) |
| Region Coverage   | 67.60% (14700/21747) |
| Branch Coverage   | 57.89% (7825/13518) |

xuchenhao commented:
run Coverage

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

gavinchou commented on be/src/common/config.cpp:
```diff
@@ -1169,7 +1169,9 @@ DEFINE_mInt64(file_cache_background_lru_dump_update_cnt_threshold, "1000");
 DEFINE_mInt64(file_cache_background_lru_log_replay_interval_ms, "1000");
 DEFINE_mBool(enable_evaluate_shadow_queue_diff, "false");
-
+DEFINE_mBool(enable_file_cache_normal_queue_2qlru, "true");
```
default value better to be false as it is not widely used.

is there any test results to show how much this mechanism improves the hit rate? 

is there any upgrade/downgrade compatibility issues if a new queue `cold_normal` and new file name `_cold_normal` introduced?

xuchenhao committed:
Enhance LRU mode compatibility.

xuchenhao committed:
Format the code.

xuchenhao committed:
Format the code.

xuchenhao committed:
Update rt cases.

xuchenhao commented:
run buildall

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao commented:
run buildall

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `79.82% (182/228)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57410_388cd4a39a15c4605b0b96dd350ab88b6821a6a2_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57410_388cd4a39a15c4605b0b96dd350ab88b6821a6a2_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 57.58% (19584/34012) |
| Line Coverage     | 43.00% (187516/436130) |
| Region Coverage   | 38.00% (150099/395030) |
| Branch Coverage   | 38.50% (63374/164624) |

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Update rt cases.

xuchenhao committed:
Update config.cpp.

xuchenhao committed:
Limit the range of file_cache_2qlru_cold_blocks_percent.

xuchenhao commented:
run buildall

xuchenhao committed:
Update config.cpp.

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57410_1587b55999c538d72bb9178f3f77c13199ec850c/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57410_1587b55999c538d72bb9178f3f77c13199ec850c/fe_report/index.html)

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/3787b5d7fb7878ccbce07ae022da61f61c6b6aa8_3787b5d7fb7878ccbce07ae022da61f61c6b6aa8_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/3787b5d7fb7878ccbce07ae022da61f61c6b6aa8_3787b5d7fb7878ccbce07ae022da61f61c6b6aa8_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 79.82% (1752/2195) |
| Line Coverage     | 65.73% (30768/46809) |
| Region Coverage   | 66.51% (15372/23112) |
| Branch Coverage   | 56.81% (8166/14374) |

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/ec3552ec76abc4f50420e7b6913dc71c230e0490_ec3552ec76abc4f50420e7b6913dc71c230e0490_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/ec3552ec76abc4f50420e7b6913dc71c230e0490_ec3552ec76abc4f50420e7b6913dc71c230e0490_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 79.72% (1757/2204) |
| Line Coverage     | 65.37% (30891/47256) |
| Region Coverage   | 66.07% (15415/23333) |
| Branch Coverage   | 56.56% (8195/14490) |

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

wenzhenghu committed:
Update comment to clarify configurable range source

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Update block_file_cache_test.cpp

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `78.95% (165/209)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/784ec0a0a50c125a9c4c22797777db56c7b92a38_784ec0a0a50c125a9c4c22797777db56c7b92a38/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/784ec0a0a50c125a9c4c22797777db56c7b92a38_784ec0a0a50c125a9c4c22797777db56c7b92a38/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 52.73% (19260/36527) |
| Line Coverage     | 36.12% (179104/495842) |
| Region Coverage   | 32.57% (138881/426401) |
| Branch Coverage   | 33.50% (60092/179387) |

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `83.73% (175/209)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57410_784ec0a0a50c125a9c4c22797777db56c7b92a38_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57410_784ec0a0a50c125a9c4c22797777db56c7b92a38_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 73.24% (26223/35805) |
| Line Coverage     | 56.27% (278310/494609) |
| Region Coverage   | 53.84% (231884/430719) |
| Branch Coverage   | 55.63% (100188/180103) |

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/84aa2acab16b8361b287212a71489c0026ed5576_84aa2acab16b8361b287212a71489c0026ed5576_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/84aa2acab16b8361b287212a71489c0026ed5576_84aa2acab16b8361b287212a71489c0026ed5576_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 79.32% (1787/2253) |
| Line Coverage     | 64.76% (31820/49139) |
| Region Coverage   | 65.46% (15859/24228) |
| Branch Coverage   | 56.02% (8427/15042) |

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao commented:
run buildall

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/33e5b24a25a062e6d649c14ac7a0a343262780c9_33e5b24a25a062e6d649c14ac7a0a343262780c9_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/33e5b24a25a062e6d649c14ac7a0a343262780c9_33e5b24a25a062e6d649c14ac7a0a343262780c9_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 79.29% (1792/2260) |
| Line Coverage     | 64.60% (31861/49323) |
| Region Coverage   | 65.36% (15889/24309) |
| Branch Coverage   | 55.88% (8438/15100) |

hello-stephen commented:
# BE UT Coverage Report
Increment line coverage `78.95% (165/209)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/33e5b24a25a062e6d649c14ac7a0a343262780c9_33e5b24a25a062e6d649c14ac7a0a343262780c9/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/33e5b24a25a062e6d649c14ac7a0a343262780c9_33e5b24a25a062e6d649c14ac7a0a343262780c9/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 52.49% (19273/36717) |
| Line Coverage     | 35.99% (179203/497959) |
| Region Coverage   | 32.41% (138987/428880) |
| Branch Coverage   | 33.34% (60112/180317) |

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `83.73% (175/209)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57410_ec8179768b598e15667bd865b2031f99dbc6711e_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57410_ec8179768b598e15667bd865b2031f99dbc6711e_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 73.20% (26352/35999) |
| Line Coverage     | 56.28% (279669/496960) |
| Region Coverage   | 53.97% (233946/433494) |
| Branch Coverage   | 55.57% (100651/181125) |

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57410_f9768440667f002d7a13a43d989265aabe4c8133/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57410_f9768440667f002d7a13a43d989265aabe4c8133/fe_report/index.html)

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao commented:
run buildall

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao commented:
run buildall

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao commented:
run buildall

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `83.73% (175/209)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57410_4737b20499460cff22fca76499c1751b9ce198bc_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57410_4737b20499460cff22fca76499c1751b9ce198bc_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 71.37% (26125/36607) |
| Line Coverage     | 54.14% (273748/505607) |
| Region Coverage   | 51.42% (227433/442277) |
| Branch Coverage   | 52.82% (97859/185258) |

xuchenhao committed:
Merge upstream/master into normal_queue_cold_hot_separation

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57410_09e2ca32214706e3d637535a6483f9e9309a018c/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57410_09e2ca32214706e3d637535a6483f9e9309a018c/fe_report/index.html)

xuchenhao commented:
run buildall

xuchenhao committed:
Merge upstream/master and resolve conflicts

xuchenhao committed:
Format the code.

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao commented:
run buildall

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57410_0cbd62a49099f33d1d9c68613c465c9f4ca6f3df/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57410_0cbd62a49099f33d1d9c68613c465c9f4ca6f3df/fe_report/index.html)

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation

xuchenhao commented:
run buildall

hello-stephen commented:
# BE Regression && UT Coverage Report
Increment line coverage `100% (0/0)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57410_21e5a80c2f6766ddd09f3d9ea98bd430090a19f4_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57410_21e5a80c2f6766ddd09f3d9ea98bd430090a19f4_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 73.30% (26997/36832) |
| Line Coverage     | 56.73% (289243/509825) |
| Region Coverage   | 54.04% (240769/445544) |
| Branch Coverage   | 55.75% (104211/186938) |

xuchenhao committed:
Merge remote-tracking branch 'upstream/master' into normal_queue_cold_hot_separation

# Conflicts:
#	be/src/io/cache/block_file_cache.cpp
#	be/test/io/cache/block_file_cache_test_lru_dump.cpp

xuchenhao commented:
run buildall

hello-stephen commented:
# Cloud UT Coverage Report
Increment line coverage ` ` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/569eb42fc91fe5540a90cc2453844b156dab1bf6_569eb42fc91fe5540a90cc2453844b156dab1bf6_cloud/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/569eb42fc91fe5540a90cc2453844b156dab1bf6_569eb42fc91fe5540a90cc2453844b156dab1bf6_cloud/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 78.34% (1906/2433) |
| Line Coverage     | 64.78% (33987/52468) |
| Region Coverage   | 65.32% (17526/26833) |
| Branch Coverage   | 53.94% (9294/17230) |

hello-stephen commented:
# FE UT Coverage Report
Increment line coverage `` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57410_569eb42fc91fe5540a90cc2453844b156dab1bf6/fe_increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57410_569eb42fc91fe5540a90cc2453844b156dab1bf6/fe_report/index.html)

xuchenhao committed:
[fix](be) Fix file cache test merge tail

### What problem does this PR solve?

Issue Number: None

Related PR: None

Problem Summary: A merge resolution on `normal_queue_cold_hot_separation` dropped the tail of `test_lru_log_record_replay_dump_restore_2qlru_to_1qlru` in `be/test/io/cache/block_file_cache_test_lru_dump.cpp`. That left the next test attached immediately after the previous assertions, which could trigger clang-format failure and break the test file structure. This commit restores the missing cleanup and function closing lines so the test file layout matches the expected style and structure.

### Release note

None

### Check List (For Author)

- Test: Manual test
    - `clang-format -i` on related BE files and `build-support/check-format.sh`
- Behavior changed: No
- Does this need documentation: No

hello-stephen commented:
# FE Regression Coverage Report
Increment line coverage `  0.00% (0/716)` :tada:
[Increment coverage report](http://coverage.selectdb-in.cc/coverage/57410_41ea6b2785a879a3bc5f49d3dda4f4149cff7317_merge_fe/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/57410_41ea6b2785a879a3bc5f49d3dda4f4149cff7317_merge_fe/report/index.html)

hello-stephen commented:
<details>
<summary>TPC-H: <b>Total hot run time: 29707 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpch-tools
Tpch sf100 test result on commit 569eb42fc91fe5540a90cc2453844b156dab1bf6, data reload: false

------ Round 1 ----------------------------------
orders	Doris	NULL	NULL	0	0	0	NULL	0	NULL	NULL	2023-12-26 18:27:23	2023-12-26 18:42:55	NULL	utf-8	NULL	NULL	
============================================
q1	17659	4088	4076	4076
q2	q3	11020	1443	827	827
q4	4802	486	345	345
q5	8606	889	596	596
q6	362	177	140	140
q7	934	857	636	636
q8	10886	1635	1648	1635
q9	7236	4554	4558	4554
q10	6814	1837	1592	1592
q11	445	275	263	263
q12	647	439	301	301
q13	18184	3437	2832	2832
q14	267	257	235	235
q15	q16	826	780	725	725
q17	901	824	969	824
q18	7034	5885	5706	5706
q19	1160	1307	1255	1255
q20	546	428	264	264
q21	5809	2827	2572	2572
q22	472	373	329	329
Total cold run time: 104610 ms
Total hot run time: 29707 ms

----- Round 2, with runtime_filter_mode=off -----
orders	Doris	NULL	NULL	150000000	42	6422171781	NULL	22778155	NULL	NULL	2023-12-26 18:27:23	2023-12-26 18:42:55	NULL	utf-8	NULL	NULL	
============================================
q1	4891	4785	5097	4785
q2	q3	4972	5265	4672	4672
q4	2068	2190	1410	1410
q5	5064	4743	4743	4743
q6	235	187	132	132
q7	1891	1757	1610	1610
q8	2296	1986	1946	1946
q9	7395	7433	7466	7433
q10	4761	4662	4224	4224
q11	545	423	365	365
q12	735	749	531	531
q13	3042	3391	2770	2770
q14	280	280	256	256
q15	q16	680	704	620	620
q17	1291	1280	1269	1269
q18	7431	6845	6831	6831
q19	1120	1116	1093	1093
q20	2235	2236	1952	1952
q21	5708	4593	4513	4513
q22	514	452	412	412
Total cold run time: 57154 ms
Total hot run time: 51567 ms
```
</details>

hello-stephen commented:
<details>
<summary>TPC-DS: <b>Total hot run time: 169922 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpcds-tools
TPC-DS sf100 test result on commit 569eb42fc91fe5540a90cc2453844b156dab1bf6, data reload: false

query5	4367	638	468	468
query6	445	206	191	191
query7	4816	597	317	317
query8	372	246	203	203
query9	8809	4094	4089	4089
query10	444	302	250	250
query11	5938	2364	2176	2176
query12	158	107	99	99
query13	1267	603	462	462
query14	6621	5426	5146	5146
query14_1	4435	4426	4430	4426
query15	214	201	182	182
query16	1059	466	485	466
query17	1141	735	615	615
query18	2736	506	356	356
query19	218	189	153	153
query20	117	110	108	108
query21	225	147	125	125
query22	13619	13605	13796	13605
query23	17404	16553	16197	16197
query23_1	16310	16313	16202	16202
query24	7590	1806	1344	1344
query24_1	1329	1326	1320	1320
query25	584	481	427	427
query26	1340	331	172	172
query27	2634	550	364	364
query28	4428	2016	2033	2016
query29	1105	660	515	515
query30	316	241	206	206
query31	1149	1082	969	969
query32	109	66	63	63
query33	563	327	270	270
query34	1178	1149	663	663
query35	785	787	719	719
query36	1385	1426	1256	1256
query37	147	107	89	89
query38	3214	3180	3075	3075
query39	950	934	892	892
query39_1	878	878	873	873
query40	218	121	99	99
query41	68	63	64	63
query42	95	96	90	90
query43	325	330	286	286
query44	
query45	197	188	186	186
query46	1123	1224	723	723
query47	2343	2345	2269	2269
query48	394	416	301	301
query49	626	468	351	351
query50	990	356	264	264
query51	4338	4313	4203	4203
query52	95	92	77	77
query53	253	274	196	196
query54	270	213	196	196
query55	79	76	71	71
query56	243	235	219	219
query57	1442	1412	1322	1322
query58	241	234	207	207
query59	1603	1654	1469	1469
query60	282	259	244	244
query61	159	158	153	153
query62	707	646	589	589
query63	230	182	185	182
query64	2502	818	611	611
query65	
query66	1762	457	346	346
query67	29840	29597	29593	29593
query68	
query69	412	317	270	270
query70	1003	987	892	892
query71	306	228	203	203
query72	3064	2740	2416	2416
query73	882	753	448	448
query74	5150	4949	4770	4770
query75	2687	2609	2243	2243
query76	2310	1179	773	773
query77	363	380	284	284
query78	12360	12369	11855	11855
query79	1423	1100	747	747
query80	730	464	399	399
query81	467	289	246	246
query82	563	160	119	119
query83	326	311	249	249
query84	260	141	113	113
query85	912	538	449	449
query86	408	298	297	297
query87	3472	3327	3192	3192
query88	3643	2746	2733	2733
query89	453	392	337	337
query90	1936	183	190	183
query91	180	168	141	141
query92	62	63	59	59
query93	1511	1455	885	885
query94	616	351	313	313
query95	679	479	348	348
query96	1115	856	338	338
query97	2693	2721	2557	2557
query98	216	207	204	204
query99	1236	1270	1035	1035
Total cold run time: 252594 ms
Total hot run time: 169922 ms
```
</details>

xuchenhao committed:
Merge remote-tracking branch 'origin/master' into pr-57410-work

# Conflicts:
#	be/src/io/cache/block_file_cache.cpp

xuchenhao committed:
[chore](be) Fix file cache replay type formatting

### What problem does this PR solve?

Issue Number: None

Related PR: #57410

Problem Summary: Adjust the formatting of the LRU replay type array in block file cache so the initializer follows the surrounding code style after the merge conflict resolution.

### Release note

None

### Check List (For Author)

- Test: No need to test (format-only change)

- Behavior changed: No

- Does this need documentation: No

xuchenhao committed:
[chore](be) Adjust file cache replay type formatting

### What problem does this PR solve?

Issue Number: None

Related PR: #57410

Problem Summary: Update the formatting of the LRU replay type array in block file cache to match the final style chosen after conflict resolution and follow-up review.

### Release note

None

### Check List (For Author)

- Test: No need to test (format-only change)

- Behavior changed: No

- Does this need documentation: No

xuchenhao committed:
[chore](be) Clean file cache replay type formatting

### What problem does this PR solve?

Issue Number: None

Related PR: #57410

Problem Summary: Remove the trailing whitespace in the LRU replay type array initializer so the final formatting matches the intended style after the merge conflict resolution.

### Release note

None

### Check List (For Author)

- Test: No need to test (format-only change)

- Behavior changed: No

- Does this need documentation: No

xuchenhao committed:
[fix](be) Fix 2Q file cache correctness gaps

### What problem does this PR solve?

Issue Number: None

Related PR: #57410

Problem Summary: Synchronize the branch with the latest master and complete COLD_NORMAL support across LRU recorder metrics, metadata deserialization and restart restoration, millisecond-based eviction timing, hot-block export filtering, and special-path eviction priority. Keep promotion state in memory while restoring normal-class blocks according to the active 2Q mode, and expose the COLD_NORMAL-to-NORMAL eviction metric with the correct destination name.

### Release note

Fix 2Q file cache correctness during recording, restart, hot-block export, promotion, eviction, and metric reporting.

### Check List (For Author)

- Test: Unit Test
    - Built BE and FE with `./build.sh --be --fe -j 24`.
    - Passed 6 focused ASAN BE unit tests covering metadata recovery, millisecond timing, eviction order, async loading, and hot-data handling.
    - Verified with an isolated FE/BE and MinIO that first access enters COLD_NORMAL, restart preserves COLD_NORMAL, and access after the promotion window moves blocks to NORMAL.
- Behavior changed: Yes. Correct COLD_NORMAL restoration, promotion, eviction, and metric behavior.
- Does this need documentation: No

xuchenhao committed:
[fix](be) Stabilize 2Q LRU dump tests

xuchenhao committed:
Merge branch 'master' into normal_queue_cold_hot_separation
